pkg/metrics/logging_hook.go

```diff
@@ -0,0 +1,85 @@
+// Copyright 2018 Authors of Cilium
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package metrics
+
+import (
+	"fmt"
+	"reflect"
+
+	"github.com/cilium/cilium/pkg/components"
+	"github.com/cilium/cilium/pkg/logging/logfields"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/sirupsen/logrus"
+)
+
+// LoggingHook is a hook for logrus which counts error, fatal, panic and warning
+// messages as a Prometheus metric.
+type LoggingHook struct {
+	component string
+}
+
+// NewLoggingHook returns a new instance of LoggingHook for the given Cilium
+// component.
+func NewLoggingHook(component string) *LoggingHook {
+	return &LoggingHook{component: component}
+}
+
+// Levels returns the list of logging levels on which the hook is triggered.
+func (h *LoggingHook) Levels() []logrus.Level {
+	return []logrus.Level{
+		logrus.PanicLevel,
+		logrus.FatalLevel,
+		logrus.ErrorLevel,
+		logrus.WarnLevel,
+	}
+}
+
+// Fire is the main method which is called every time when logger has an error,
+// fatal, panic or warning message.
+func (h *LoggingHook) Fire(entry *logrus.Entry) error {
+	var metric prometheus.CounterVec
+
+	// NOTE(mrostecki): For now errors and warning metric exists only for Cilium
+	// daemon, but support of Prometheus metrics in some other components (i.e.
+	// cilium-health - GH-4268) is planned.
+
+	// Pick a metric for the component.
+	switch h.component {
+	case components.CiliumAgentName:
+		metric = ErrorsWarnings
```

Comment:
This is something that can be part of LoggingHook as well.

This method (`Fire(entry *logrus.Entry)`) will be called for all messages so it needs to be short and fast as possible.